sweep/fee_bumper.go

```diff
@@ -0,0 +1,623 @@
+	// for new blocks.
+	chainNotifier chainntnfs.ChainNotifier
+
+	// currentHeight stores the currently best known block height.
+	currentHeight uint32
+
+	// monitoredInputs is a map that stores the requested and unconfirmed
+	// UTXOs.
+	monitoredInputs lnutils.SyncMap[*wire.OutPoint, *monitoredInput]
+
+	// updateParams mounts the function `UpdateParams` of the sweeper so we
+	// can easily send a request to update the fee rate for a given
+	// input.
+	updateParams paramsUpdater
+
+	changeOutputScript []byte
+
+	// quit is closed when the fee bumper stops.
+	quit chan struct{}
+}
+
+// A compile time check to ensure feeBumper implements Bumper interface.
+var _ Bumper = (*feeBumper)(nil)
+
+// newFeeBumper creates a new feeBumper.
+func newFeeBumper(chainNotifier chainntnfs.ChainNotifier,
+	updateParams paramsUpdater, pkScript []byte) *feeBumper {
+
+	return &feeBumper{
+		chainNotifier: chainNotifier,
+		updateParams:  updateParams,
+		monitoredInputs: lnutils.SyncMap[
+			*wire.OutPoint, *monitoredInput,
+		]{},
+		changeOutputScript: pkScript,
+		quit:               make(chan struct{}),
+	}
+}
+
+// NotifyConfirm tells the bumper a transaction has been confirmed and needs to
+// be removed from the monitor queue.
+//
+// NOTE: part of the Bumper interface.
+func (f *feeBumper) NotifyConfirm(op *wire.OutPoint) {
+	input, ok := f.monitoredInputs.LoadAndDelete(op)
+	if !ok {
+		feeLog.Infof("Notify the spend of outpoint: %v not found in "+
+			"monitored txes, might have been deleted", op)
+		return
+	}
+
+	feeLog.Infof("Notified the spent: outpoint=%v deleted", op)
+
+	feeLog.Tracef("Outpoint confirmed: %s", input)
+}
+
+// validateRequest checks that the request has sane params used.
+func (f *feeBumper) validateRequest(req *MonitorRequest) error {
+	// Validate the strategy used.
+	if err := validateStrategy(req.Strategy); err != nil {
+		return err
+	}
+
+	// Validate conf target and deadline height.
+	if err := validateConfTarget(req); err != nil {
+		return err
+	}
+
+	feeLog.Debugf("Validated request for outpoint: %v",
+		req.Input.OutPoint())
+
+	return nil
+}
+
+// validateConfTarget validates the conf target is sane.
+func validateConfTarget(req *MonitorRequest) error {
+	// Check conf target is greater than 0.
+	if req.ConfTarget < 1 {
+		return ErrConfTargetTooSmall
+	}
+
+	if req.ConfTarget == NoHeightInfo && req.FeeRate == 0 {
+		return fmt.Errorf("%w: must set conf target or fee rate",
+			ErrConflictArgs)
+
+	}
+
+	return nil
+}
+
+// RequestMonitor registers an outpoint to be monitored for potential fee
+// bumping. If the outpoint already exits, ErrAlreadyMonitored will be
+// returned.
+//
+// NOTE: part of the Bumper interface.
+func (f *feeBumper) RequestMonitor(req *MonitorRequest) error {
+	feeLog.Debugf("Received request %s", req)
+
+	op := req.Input.OutPoint()
+
+	// Check if the request has already been processed.
+	//
+	// TODO(yy): allow overwrite?
```

Comment:
Or return a `nil` error unless we expect the caller to properly handle the returned `ErrAlreadyMonitored`.